### PR TITLE
fix altercolumns db migration

### LINKED ISSUE #8560: columnsToAlter on strapi-database-schema is no awaited before use
Hello!
I think I found a possible bug, or at least unreachable code.

getColumnsWhereDefinitionChanged is an async function, that always return a promise.
On strapi-connector-bookshelf/lib/build-database-schema.js the function is called without await.
So, the length of columnsToAlter  will be always undefined.
https://github.com/strapi/strapi/blob/3c78413d9ce3ffd753c37837c4d08e60fe74b63a/packages/strapi-connector-bookshelf/lib/build-database-schema.js#L305-L311

image
image

I could make a PR, but I prefer someone on the team check that the code we are going to make reachable is safe (beacuse contains DROP)

Thanks!

### PR DESCRIPTION
**Pull request by petersg83**
fix #8560

### CODE REVIEW

**Review by alexandrebodin — COMMENTED** [review]
LGTM, waiting for the test fix before approving

File: packages/strapi-connector-bookshelf/lib/migrations/index.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
+'use strict';
+
+const dpMigration = require('./draft-publish-migration');
+
+const migrations = [dpMigration];
```
Comment (alexandrebodin) [inline packages/strapi-connector-bookshelf/lib/migrations/index.js:5]: good idea. We could push it a bit further by creating a `migrationRunner` that will wrap those lifecycles logic arroudn the schemaMigration

**Review by alexandrebodin — COMMENTED** [review]
Waiting on your response on the comment before checking the rest

File: packages/strapi-connector-bookshelf/lib/migrations/index.js (hunk 1 above)
Comment (petersg83) [reply packages/strapi-connector-bookshelf/lib/migrations/index.js:5]: something like that :
```
await migrationRunner(createOrUpdateTables)({ ORM, loadedModel, definition, connection, model });
```
?

File: packages/strapi-connector-bookshelf/lib/build-database-schema.js [hunk 2]
```diff
@@ -390,3 +380,15 @@ const createOrUpdateTable = async ({ table, attributes, definition, ORM, model }
     }
   }
 };
+
+const migrationRunner = createMigrationRunner(migrateSchemas, {
+  migrations: [draftPublishMigration],
```
Comment (alexandrebodin) [inline packages/strapi-connector-bookshelf/lib/build-database-schema.js:385]: Your param here should be `hooks` ?

**Review by alexandrebodin — CHANGES_REQUESTED** [review]

File: packages/strapi-database/test/migration-draft-publish.test.e2e.js [hunk 3]
```diff
@@ -35,88 +35,95 @@ const dogs = [
...
+      test('No published_at before enabling the feature', async () => {
+        let { body } = await rq({
+          url: '/content-manager/explorer/application::dog.dog',
+          method: 'GET',
+        });
+        expect(body.length).toBe(2);
+        const expectDog = expectDog => dog =>
+          dog.name === expectDog.name && _.isUndefined(dog.published_at);
+        expect(body.find(expectDog(data.dogs[0])));
+        expect(body.find(expectDog(data.dogs[1])));
```
Comment (petersg83) [inline packages/strapi-database/test/migration-draft-publish.test.e2e.js:76]: doesn't work

**Review by alexandrebodin — APPROVED** [review]
LGTM 👍

**Pull request merged**